File: Medium/add_two_numbers.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	node := &ListNode{}
	current := node
	carry := 0

	for l1 != nil || l2 != nil || carry > 0 {
		v1, v2 := 0, 0

		if l1 != nil {
			v1 = l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			v2 = l2.Val
			l2 = l2.Next
		}

		sum := v1 + v2 + carry
		carry = sum / 10

		current.Next = &ListNode{Val: sum % 10}
		current = current.Next
	}

	return node.Next
}

func arrayToList(arr []int) *ListNode {
	node := &ListNode{}
	current := node

	for _, val := range arr {
		current.Next = &ListNode{Val: val}
		current = current.Next
	}

	return node.Next
}

func listToArray(node *ListNode) []int {
	var result []int
	for node != nil {
		result = append(result, node.Val)
		node = node.Next
	}
	return result
}

func run_addTwoNumbers() {
	listTest := [][][]int{
		{
			{2, 4, 3}, // First number
			{5, 6, 4}, // Second number
		},
		{
			{9, 9, 9, 9, 9, 9, 9},
			{9, 9, 9, 9},
		},
		{
			{0},
			{0},
		},
	}

	for _, num := range listTest {
		num1 := arrayToList(num[0])
		num2 := arrayToList(num[1])
		fmt.Println(listToArray(addTwoNumbers(num1, num2)))
	}
}

File: Medium/add_two_numbers.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        node = ListNode()
        current = node
        carry = 0

        while l1 or l2 or carry:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0

            s = val1 + val2 + carry
            carry = s // 10
            current.next = ListNode(s % 10)
            current = current.next

            if l1:
                l1 = l1.next

            if l2:
                l2 = l2.next

        return node.next
    
print(Solution().addTwoNumbers([2, 4, 3], [5, 6, 4]))
print(Solution().addTwoNumbers([9, 9, 9, 9, 9, 9, 9], [9, 9, 9, 9]))
print(Solution().addTwoNumbers([0], [0]))